coins/monero/src/wallet/seed/classic/mod.rs

```diff
@@ -0,0 +1,250 @@
+use std::collections::HashMap;
+
+use crc::{Crc, CRC_32_ISO_HDLC};
+use rand_core::OsRng;
+
+use crate::random_scalar;
+
+use super::{SeedError, LanguageName};
+
+mod languages;
+
+pub const CLASSIC_SEED_LENGTH: usize = 24;
+pub const CLASSIC_SEED_LENGTH_WITH_CHECKSUM: usize = 25;
+const LANGUAGE_WORD_COUNT: usize = 1626;
+
+#[derive(Clone)]
+pub struct Language {
+  word_list: Vec<String>,
+  word_map: HashMap<String, usize>,
+  trimmed_word_map: HashMap<String, usize>,
+  language_name: LanguageName,
+  unique_prefix_length: usize,
+}
+
+impl Language {
+  /// populates the word_map and trimmed_word_maps
+  fn populate_maps(
+    &mut self,
+    allow_short_words: bool,
+    allow_duplicate_prefix: bool,
+  ) -> Result<(), SeedError> {
+    if self.word_list.len() != LANGUAGE_WORD_COUNT {
+      Err(SeedError::InvalidLanguageWordList)?;
+    }
+
+    for (i, word) in self.word_list.iter().enumerate() {
+      self.word_map.insert(word.clone(), i);
+
+      if word.len() < self.unique_prefix_length && !allow_short_words {
+        Err(SeedError::InvalidWordInLanguage)?;
+      }
+
+      let trimmed = if word.len() > self.unique_prefix_length {
+        utf8_prefix(word, self.unique_prefix_length)
+      } else {
+        word.to_string()
+      };
+
+      if self.trimmed_word_map.contains_key(&trimmed) && !allow_duplicate_prefix {
+        Err(SeedError::DuplicatePrefix)?;
+      }
+
+      self.trimmed_word_map.insert(trimmed, i);
```

Comment:
This will override the previous value and make it impossible to properly use trimmed seeds. Is that intended behavior? For these languages to not support trimmed seeds?

If so, they should have None for trimmed_word_map, not an invalid map.

Also, since this is solely internal, this may be re-writable as something like
```rs
if self.trimmed_word_map.contains_key(&trimmed) {
  self.trimmed_word_map = None;
}
```

Then we can use tests to ensure languages which should support trimmed seeds do in fact support trimmed seeds.